### DISCUSSION TITLE
At the mercy of the machines

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12261] [Using MicroPython] [answered]
A year or so back I did a pip.install mpy-cross==1.17 on windows 10. The machine elected to park the result in C:\Users\Admin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\mpy_cross
It took me ages to find but once I did it worked great at squashing .py files for the esp32.

Now days I've got more & more esp32s running ver 1.20 & I'd like make some .mpy files for them too.  I'm worried that pip.install mpy-cross==1.20 is going clobber the existing mpy_cross for v1.17 which I'm still using. 

Is there some cmd line switch I can invoke that will force the pip install to put the new mpy_cross somewhere else?

**Comment by felixdoerre** [comment, marked as answer]
You could use a virtualenv to install a package "locally" into that virtualenv (https://docs.python.org/3/library/venv.html). That will not touch system-wide installed packages.

**Reply by jimmo** [reply]
+1 to this -- this is definitely the way to work with Python dependencies.

FYI we are also looking at solving this in teh future by having the mpy_cross PyPI package support multiple bytecode versions. See https://github.com/micropython/micropython/pull/11723 for the initial work.

**Comment by dlech** [comment]
At Pybricks, we needed to be able to install multiple versions of `mpy-cross` at the same time, so we made alternate `mpy-cross-v5` and `mpy-cross-v6` packages that work according to the compatibility table at https://docs.micropython.org/en/latest/reference/mpyfiles.html#versioning-and-compatibility-of-mpy-files

**Comment by kjm1102** [comment]
Mpy.cross is one of the few python 'extras' simple enough for the likes of
me Jim, hope it's still a simple install when you've finished expanding its
scope. Looking forward to it because the other 2 suggestions are completely
beyond me.
